### PR TITLE
fix(gradle-wrapper): support for wrapper updates in sub-directory

### LINKED ISSUE #14371: Gradle-Wrapper not complete for projects in sub-directories
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

When a Gradle project is in a sub-directory of a repository, the "basic" update of the `gradle-wrapper.properties` works, but related changes (SHA256 sum, script updates) are not applied. This happens in a self-hosted Renovate installation, but I build a reproducer project here: https://github.com/TobiX/test-gradle-renovate

This might be related to #13081 and #5377

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: PR rebase requested=true(branch="renovate/gradle-7.x")
DEBUG: Checking if PR has been edited(branch="renovate/gradle-7.x")
DEBUG: Branch has not been modified(branch="renovate/gradle-7.x")
{
  "branchName": "renovate/gradle-7.x"
}
DEBUG: Found existing branch PR(branch="renovate/gradle-7.x")
DEBUG: Checking schedule(at any time, null)(branch="renovate/gradle-7.x")
DEBUG: No schedule defined(branch="renovate/gradle-7.x")
DEBUG: Manual rebase requested via Dependency Dashboard(branch="renovate/gradle-7.x")
DEBUG: Using reuseExistingBranch: false(branch="renovate/gradle-7.x")
DEBUG: manager.getUpdatedPackageFiles() reuseExistinbranch=false(branch="renovate/gradle-7.x")
DEBUG: Starting search at index 168(packageFile="demo/gradle/wrapper/gradle-wrapper.properties", branch="renovate/gradle-7.x")
{
  "depName": "gradle"
}
DEBUG: Found match at index 168(packageFile="demo/gradle/wrapper/gradle-wrapper.properties", branch="renovate/gradle-7.x")
{
  "depName": "gradle"
}
DEBUG: Contents updated(packageFile="demo/gradle/wrapper/gradle-wrapper.properties", branch="renovate/gradle-7.x")
{
  "depName": "gradle"
}
DEBUG: gradle-wrapper.updateArtifacts()(branch="renovate/gradle-7.x")
[...]
INFO: No gradlew found - skipping Artifacts update(branch="renovate/gradle-7.x")
DEBUG: Updated 1 package files(branch="renovate/gradle-7.x")
DEBUG: No updated lock files in branch(branch="renovate/gradle-7.x")
DEBUG: isBranchConflicted(main, renovate/gradle-7.x)(branch="renovate/gradle-7.x")
DEBUG: 1 file(s) to commit(branch="renovate/gradle-7.x")
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### LINKED ISSUE #17057: Gradle wrapper distributionSha256Sum still not updating
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

_No response_

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Updates to gradle wrapper versions in `gradle-wrapper.properties` which contain a `distributionSha256Sum` still appear to not be updating the checksum, and only bumping the version.

Previous issues #6148 and #13081 seem to indicate that this has been a problem before and was fixed, so not sure if I'm running into a regression or a new bug manifest in a similar way.

See the PR at https://github.com/owntracks/android/pull/1264/files for an example of where renovate is bumping `distributionUrl` but not changing `distributionSha256Sum`.

### Relevant debug logs

<details><summary>Logs</summary>

```
 "gradle-wrapper": [
      {
        "packageFile": "project/gradle/wrapper/gradle-wrapper.properties",
        "deps": [
          {
            "depName": "gradle",
            "currentValue": "7.5",
            "replaceString": "https\\://services.gradle.org/distributions/gradle-7.5-all.zip",
            "datasource": "gradle-version",
            "versioning": "gradle",
            "depIndex": 0,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "7.5.1",
                "newValue": "7.5.1",
                "releaseTimestamp": "2022-08-05T21:17:56.000Z",
                "newMajor": 7,
                "newMinor": 5,
                "updateType": "patch",
                "branchName": "renovate/gradle-7.x"
              }
            ],
            "warnings": [],
            "sourceUrl": "https://github.com/gradle/gradle",
            "homepage": "https://gradle.org",
            "currentVersion": "7.5",
            "isSingleVersion": true,
            "fixedVersion": "7.5"
          }
        ]
      }
    ]
  }
}
```

</details>

Logs from https://app.renovatebot.com/dashboard#github/owntracks/android/767100327

### Have you created a minimal reproduction repository?

No reproduction, but I have linked to a public repo where it occurs

### PR DESCRIPTION
**Pull request by Churro**
## Changes

- Adds support for gradle-wrapper updates, if gradle-wrapper is located in a sub-directory
- Updates `distributionSha256Sum` and, if the gradle build runs through successfully, wrapper artifacts too

## Context

- Closes https://github.com/renovatebot/renovate/issues/14371
  - Test repo: https://github.com/Churro/renovate-gradle-wrapper/pull/1

- Closes https://github.com/renovatebot/renovate/issues/17057
  - Test repo: https://github.com/Churro/owntracks-android/pull/1
  - _Note_: the PR misses the wrapper artifacts as I don't have the Android SDK installed and gradle fails with: 
   > SDK location not found. Define a valid SDK location with an ANDROID_HOME environment variable or by setting the sdk.dir path in your project's local properties file at ...
     - If you run renovate on a host with correctly setup Android SDK, the PR should also include updated artifacts. I guess this doesn't apply to the Github App though.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/gradle-wrapper/artifacts.spec.ts [hunk 1]
```diff
@@ -187,10 +187,46 @@ describe('modules/manager/gradle-wrapper/artifacts', () => {
...
+        'sub/gradlew.bat',
+      ].map((fileProjectPath) => ({
+        file: {
+          type: 'addition',
+          path: fileProjectPath,
+          contents: 'test',
+        },
+      }))
+    );
+    expect(execSnapshots).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/manager/gradle-wrapper/artifacts.spec.ts:230]: don't use snapshots for new tests, read our best practices guide in docs/development

File: lib/modules/manager/gradle-wrapper/utils.ts [hunk 2]
```diff
@@ -23,22 +23,21 @@ export function gradleWrapperFileName(): string {
 }
 
 export async function prepareGradleCommand(
-  gradlewName: string,
+  gradlewFile: string,
```
Comment (viceice) [inline lib/modules/manager/gradle-wrapper/utils.ts:26]: don't do unrelated changes

File: lib/modules/manager/gradle-wrapper/artifacts.ts [hunk 3]
```diff
@@ -84,6 +87,8 @@ export async function updateArtifacts({
     } else {
       cmd += ` --gradle-version ${quote(config.newValue!)}`;
     }
+    cmd += ` --project-dir ${localGradleDir}`;
```
Comment (viceice) [inline lib/modules/manager/gradle-wrapper/artifacts.ts:90]: i think you should instead pass one of the cwd exec options like all other managers do.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/gradle-wrapper/utils.ts (hunk 2 above)
Comment (Churro) [reply lib/modules/manager/gradle-wrapper/utils.ts:26]: this change is related

File: lib/modules/manager/gradle-wrapper/artifacts.spec.ts [hunk 4]
```diff
@@ -187,10 +187,52 @@ describe('modules/manager/gradle-wrapper/artifacts', () => {
...
+        file: {
+          type: 'addition',
+          path: fileProjectPath,
+          contents: 'test',
+        },
+      }))
+    );
+    expect(execSnapshots).toMatchObject([
+      {
+        options: {
```
Comment (viceice) [inline lib/modules/manager/gradle-wrapper/artifacts.spec.ts:232]:
```suggestion
        cmd: '<gradlew> wrapper --gradle-version 5.6.4',
        options: {
```

we need to make sure it's called with right path

**Review by viceice — CHANGES_REQUESTED** [review]
- blocked by #17204

**Comment by viceice** [comment]
needs deconflicting now 👼

File: lib/modules/manager/gradle-wrapper/artifacts.spec.ts [hunk 5]
```diff
@@ -275,4 +275,51 @@ describe('modules/manager/gradle-wrapper/artifacts', () => {
...
+
+  it('handles gradle-wrapper in subdirectory', async () => {
+    const execSnapshots = mockExecAll();
+    git.getRepoStatus.mockResolvedValue({
+      modified: [
+        'sub/gradle/wrapper/gradle-wrapper.properties',
+        'sub/gradlew',
+        'sub/gradlew.bat',
+      ],
+    } as StatusResult);
```
Comment (viceice) [inline lib/modules/manager/gradle-wrapper/artifacts.spec.ts:287]: please use partial helper like other tests

File: lib/modules/manager/gradle-wrapper/artifacts.spec.ts [hunk 6]
```diff
@@ -275,4 +275,51 @@ describe('modules/manager/gradle-wrapper/artifacts', () => {
...
+    const res = await updateArtifacts({
+      packageFileName: 'sub/gradle/wrapper/gradle-wrapper.properties',
+      updatedDeps: [],
+      newPackageFileContent: Fixtures.get(
+        'expectedFiles/gradle/wrapper/gradle-wrapper.properties'
+      ),
+      config: { ...config, newValue: '6.3' },
+    });
+
+    expect(res).toEqual(
```
Comment (viceice) [inline lib/modules/manager/gradle-wrapper/artifacts.spec.ts:298]: don't use map, check other tests

File: lib/modules/manager/gradle-wrapper/utils.ts [hunk 7]
```diff
@@ -21,16 +21,17 @@ export function gradleWrapperFileName(): string {
   return './gradlew';
 }
 
-export async function prepareGradleCommand(): Promise<string | null> {
-  const gradlewFile = gradleWrapperFileName();
+export async function prepareGradleCommand(
```
Comment (viceice) [inline lib/modules/manager/gradle-wrapper/utils.ts:24]: looks like a wrong merge, we don't need he argument, but maybe we need the relative path

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/gradle-wrapper/artifacts.spec.ts (hunk 6 above)
Comment (Churro) [reply lib/modules/manager/gradle-wrapper/artifacts.spec.ts:298]: I copied this 1:1 from your refactored tests! 

https://github.com/renovatebot/renovate/blob/a657c3476e7a6c78034147ae04c71ae81b88bd17/lib/modules/manager/gradle-wrapper/artifacts.spec.ts#L93-L100

File: lib/modules/manager/gradle-wrapper/utils.ts (hunk 7 above)
Comment (Churro) [reply lib/modules/manager/gradle-wrapper/utils.ts:24]: For `statLocalFile()` to work on a subfolder, the path to `gradlew` in that subfolder is needed.

File: lib/modules/manager/gradle-wrapper/artifacts.spec.ts (hunk 6 above)
Comment (Churro) [reply lib/modules/manager/gradle-wrapper/artifacts.spec.ts:300]: Apart from that, I don't see the problem with map

Comment (viceice) [reply lib/modules/manager/gradle-wrapper/artifacts.spec.ts:300]: ok, forgot to refactor that test 😅

**Review by viceice — APPROVED** [review]
